packages/lesswrong/server/resolvers/moderationResolvers.ts

```diff
@@ -30,3 +35,91 @@ defineQuery({
     };
   }
 });
+
+defineMutation({
+  name: "lockThread",
+  argTypes: "(commentId: String!, until: String)",
+  resultType: "Boolean!",
+  fn: async (_root: void, args: {commentId: string, until?: string}, context: ResolverContext) => {
+    const { currentUser } = context;
+    if (!userIsAdminOrMod(currentUser)) {
+      throw new Error("Only admins and moderators can lock or unlock threads");
+    }
+    
+    // Get a list of all comments nested under this one, recursively
+    const rootComment = await Comments.findOne({_id: args.commentId});
+    if (!rootComment) {
+      throw new Error("Invalid comment ID for lockThread");
+    }
+    const commentsInThread: DbComment[] = await getCommentSubtree(rootComment);
+    
+    const farFuture = moment().add(1000, 'years').toDate();
+    const expiryDate = args.until
+      ? moment(args.until).toDate()
+      : farFuture;
+    
+    // Mark them all as replies-locked
+    await Promise.all(commentsInThread.map(async (comment) => {
```

Comment:
This seems a bit scary if we lock a comment with a lot of descendants.  I guess even if we wanted to do a raw batch update we'd still want to run callbacks for all of the affected comments, which will still cause a lot of concurrent db queries.